### PR TITLE
Handling tenancy related edge cases

### LINKED ISSUE #990: [Tenant] Inconsistent docker volume ls output in case of orphaned VMs
Do not see consistent output for 'docker volume ls' command across vms after _DEFAULT tenant is deleted. 

Setup:
One ESX, 
Two VMs (TestVM1-125 and TestVM2-125)
VSAN datastore.
Both VMs belong to same vsan datastore and same esx.

Steps:
1. Created a docker volume named testVolVsan from vm2  (TestVM2-125).
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume create --driver=vsphere --name=testVolVsan -o size=200mb
testVolVsan
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             testVolVsan@vsanDatastore
vsphere             vol1@sharedVmfs-0
vsphere             vol2@sharedVmfs-0
```
Docker volume ls command from VM1
```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
DRIVER              VOLUME NAME
vsphere             testVolVsan@vsanDatastore
vsphere             vol1@sharedVmfs-0
vsphere             vol2@sharedVmfs-0
```

2. Admin cli command output listing volumes and default tenant .
```
[root@sc-rdops-vm18-dhcp-55-125:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls
Volume       Datastore      Created By VM  Created                   Attached To VM (name/uuid)  Policy          Capacity  Used   Disk Format  Filesystem Type  Access      Attach As
-----------  -------------  -------------  ------------------------  --------------------------  --------------  --------  -----  -----------  ---------------  ----------  ----------------------
vol1         sharedVmfs-0   ubuntu-VM0.0   Fri Mar  3 19:21:15 2017  detached                    N/A             10GB      145MB  thin         ext4             read-write  independent_persistent
vol2         sharedVmfs-0   ubuntu-VM1.0   Fri Mar  3 19:21:34 2017  detached                    N/A             10GB      145MB  thin         ext4             read-write  independent_persistent
testVolVsan  vsanDatastore  TestVM2-125    Fri Mar  3 19:55:00 2017  detached                    [VSAN default]  200MB     96MB   thin         ext4             read-write  independent_persistent

[root@sc-rdops-vm18-dhcp-55-125:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant ls
Uuid                                  Name      Description               Default_datastore  VM_list
------------------------------------  --------  ------------------------  -----------------  -------
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default tenant
```
3. Then deleted the default tenant. Do not see any volumes or default tenant
```
[root@sc-rdops-vm18-dhcp-55-125:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant rm --name=_DEFAULT
tenant rm succeeded
[root@sc-rdops-vm18-dhcp-55-125:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py tenant ls
Uuid  Name  Description  Default_datastore  VM_list
----  ----  -----------  -----------------  -------


[root@sc-rdops-vm18-dhcp-55-125:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py ls
Volume  Datastore  Created By VM  Created  Attached To VM (name/uuid)  Policy  Capacity  Used  Disk Format  Filesystem Type  Access  Attach As
------  ---------  -------------  -------  --------------------------  ------  --------  ----  -----------  ---------------  ------  ---------

```
4. Do not see same output for for docker volume ls command.

'docker volume ls' command from VM1

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
list vsphere: VolumeDriver.List: VM TestVM1-125 does not belong to any tenant
DRIVER              VOLUME NAME
```

'docker volume ls' command from VM2

```
root@sc-rdops-vm02-dhcp-52-237:~# docker volume ls
list vsphere: VolumeDriver.List: VM TestVM2-125 does not belong to any tenant
DRIVER              VOLUME NAME
vsphere             testVolVsan
vsphere             testVolVsan@vsanDatastore
root@sc-rdops-vm02-dhcp-52-237:~#
```

Logs:
[docker-volume-vsphere-VM2.txt](https://github.com/vmware/docker-volume-vsphere/files/818099/docker-volume-vsphere-VM2.txt)
[vmdk_ops.txt](https://github.com/vmware/docker-volume-vsphere/files/818101/vmdk_ops.txt)
[docker-volume-vsphere-VM1.txt](https://github.com/vmware/docker-volume-vsphere/files/818100/docker-volume-vsphere-VM1.txt)

### LINKED ISSUE #1045: [Tenancy] Volume attach/detach should handle inflight membership change of VM to tenant
Change of membership(which tenant a VM belongs to) after a volume as been attached but before detach makes the volume detach logic fail. The reason being tenant not found (if tenant is deleted and even default tenant doesn't exist) or vmdk not found if tenant is changed(vmdk is now searched in new tenant uuid directory). 

The fix is to only match datastore path and vmdk name and ignore tenant name against the volumes that has been attached to the VM to allow smooth detach. 

There can be a theoretical corner case where volumes of same name exist across 2 (or may be more) tenants, and while a container is running and say vol1 from tenant 1 is attached; and admin moves the VM from tenant 1 to tenant 2 and tries to mount vol1 from tenant 2 to the same container. 
In this case, the above detach logic will fail.
To avoid this we restrict the attach to allow unique volume names to be mounted. This means, in the above example, the second mount (vol1 from tenant 2) will fail; of course with appropriate message.

### PR DESCRIPTION
**Pull request by pshahzeb**
1. Disallow tenant rm, tenant vm rm, tenant vm replace when VMs which are part of the given tenant have any volumes currently mounted and being used. Safe approach to alert the user so that he can complete the ongoing jobs and then manipulate the tenancy membership changes.

2. Send empty list as output to docker volume ls command to forcefully make docker
   print emtpy list and not the cached volume names

3. [Minor] print to indicate negative test case and avoid CI output confusion

Fixes #1045 
Create tenant1 and add a vm to it.
```
[root@sc-rdops-vm05-dhcp-133-3:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name      Description               Default_datastore  VM_list
------------------------------------  --------  ------------------------  -----------------  ------------
11111111-1111-1111-1111-111111111111  _DEFAULT  This is a default tenant
b06c6101-6b01-44c5-a318-09bb85392fd9  tenant1                                                photon-VM0.1
```

Create a volume and attach it to a container inside this VM. Keep the container running.
```
root@photon-xVmYMbyTn [ ~ ]# docker volume create --driver=vsphere --name  myVolume
myVolume
root@photon-xVmYMbyTn [ ~ ]# docker run -it --volume-driver=vsphere -v myVolume@sharedVmfs-0:/vol1 --name ub ubuntu
root@ada941b998cf:/#
```

Try to remove the VM from tenant, replace the VM with another VM and try to delete the tenant.
Error with appropriate message
```
[root@sc-rdops-vm05-dhcp-133-3:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group vm rm --name tenant1 --vm-list photon-VM0.1
Cannot complete vm-group vm rm. VM 'photon-VM0.1' has volumes mounted.
[root@sc-rdops-vm05-dhcp-133-3:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group vm replace --name tenant1 --vm-list photon-VM1.1
Cannot complete vm-group vm replace. VM 'photon-VM0.1' has volumes mounted.
[root@sc-rdops-vm05-dhcp-133-3:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group rm --name tenant1
Cannot complete tenant rm. VM 'photon-VM0.1' has volumes mounted.
```

VM is orphan now. docker volume ls shows empty list of volumes
```
root@photon-xVmYMbyTn [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
```

Fixes #990 
Create a volume from a VM and then delete the VM's tenant
```
root@photon-xVmYMbyTn [ ~ ]# docker volume create --driver=vsphere --name i_will_be_orphan
i_will_be_orphan
root@photon-xVmYMbyTn [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
vsphere             i_will_be_orphan@sharedVmfs-0
```

```
[root@sc-rdops-vm05-dhcp-133-3:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group ls
Uuid                                  Name     Description  Default_datastore  VM_list
------------------------------------  -------  -----------  -----------------  ------------
b06c6101-6b01-44c5-a318-09bb85392fd9  tenant1                                  photon-VM0.1

[root@sc-rdops-vm05-dhcp-133-3:~] /usr/lib/vmware/vmdkops/bin/vmdkops_admin.py vm-group rm --name tenant1
vm-group rm succeeded
```

docker volume ls from the VM gives empty list
```
root@photon-xVmYMbyTn [ ~ ]# docker volume ls
DRIVER              VOLUME NAME
```

### CODE REVIEW

File: esx_service/vmdk_ops.py [hunk 1]
```diff
@@ -786,9 +786,23 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     vm_datastore_url = vmdk_utils.get_datastore_url_from_config_path(config_path)
     vm_datastore = get_datastore_name(vm_datastore_url)
 
+    # Detach is implicity allowed if a disk has been attached
+    # Do not find tenant or privileges; just detach the disk
+    if cmd == "detach":
```
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: Can we introduce an enum to define all the commands?

File: esx_service/vmdk_ops.py [hunk 2]
```diff
@@ -656,11 +656,11 @@ def attachVMDK(vmdk_path, vm_uuid):
 
 
 # Return error, or None for OK.
-def detachVMDK(vmdk_path, vm_uuid):
+def detachVMDK(full_vol_name, vm_uuid):
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: I prefer full_vmdk_path rather than full_vol_name.

File: esx_service/vmdk_ops.py [hunk 3]
```diff
@@ -786,9 +786,23 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
...
+            response = detachVMDK(full_vol_name, vm_uuid)
+
+        return response
+
     error_info, tenant_uuid, tenant_name = auth.get_tenant(vm_uuid)
     if error_info:
-        return err(error_info)
+        # For docker volume ls, docker prints a list of cached volume names in case
+        # of error(in this case, orphan VM).
+        # Explicity providing empty list of volumes to avoid misleading output.
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: How do you differentiate between genuine error versus VM not member of any tenant?

Please add link to issue.

File: esx_service/vmdk_ops.py [hunk 4]
```diff
@@ -1297,16 +1311,20 @@ def err(string):
     return {u'Error': string}
 
 
-def disk_detach(vmdk_path, vm):
-    """detach disk (by full path) from a vm amd return None or err(msg)"""
+def disk_detach(full_vol_name, vm):
```
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: Same here. Since we are manipulating VMDK, I would refer this as full_vmdk_path

Comment (pshahzeb) [inline esx_service/vmdk_ops.py]: But the content of param 'full_vol_name' is qualified volume name in format volume@datastore
Calling it full_vmdk_path would be confusing.
I will update the param comment though

File: esx_service/vmdk_ops.py (hunk 2 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: so is it  a name (vol@DS) or a path? 
A comment (in a dockstring) would be good

File: esx_service/vmdk_ops.py (hunk 1 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: Good suggestion, but outside of scope for this PR. @shaominchen  - Please file an issue (or code it up and submit a PR) 

Generally, executeRequest() became too messy. It was a small function with a switch for 5 simple commands, now it's quite complex. Would be nice to clean up. Certainly not in this PR

File: esx_service/vmdk_ops.py [hunk 5]
```diff
@@ -786,9 +786,23 @@ def executeRequest(vm_uuid, vm_name, config_path, cmd, full_vol_name, opts):
     vm_datastore_url = vmdk_utils.get_datastore_url_from_config_path(config_path)
     vm_datastore = get_datastore_name(vm_datastore_url)
 
+    # Detach is implicity allowed if a disk has been attached
+    # Do not find tenant or privileges; just detach the disk
+    if cmd == "detach":
+        with lockManager.get_lock(vm_uuid):
+            response = detachVMDK(full_vol_name, vm_uuid)
+
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: nit (can be ignored): generally I'd prefer to have a simple dict-driven code, with something like  {"attach": ["auth_needed": True, "func": "code", ,,} ...}  to drive the execution for commands. (this is the same note as the above reply to @shaominchen ).

File: esx_service/vmdk_ops.py (hunk 3 above)
Comment (msterin) [inline esx_service/vmdk_ops.py]: I'd suggest using `if err_info and  tenant_uuid: <read_error>`  and `if erron_info and !tenant_uuid` - no tenant

But yes, this one hides all real errors from get_tenant, if any. (mainly DB errors/file IO I guess)

File: esx_service/vmdk_ops.py [hunk 6]
```diff
@@ -950,20 +965,18 @@ def findDeviceByPath(vmdk_path, vm):
...
-        if virtual_disk == backing_disk:
-            logging.debug("findDeviceByPath: MATCH: %s", backing_disk)
-            return d
-    return None
+        backing_disk_name = vmdk_utils.strip_vmdk_extension(os.path.basename(backing_disk))
+
+        logging.debug("findDeviceByName: datastore = %s, backing_disk_name = %s",
+                        datastore, backing_disk_name)
+
+        if backing_disk_name == vol_name:
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: So what if there are 2 Docker volumes with the same names on different datastores. both attached ? Which one will be detached?  I do not see DS names used in matching.... not 'dockvols' part of the path

File: esx_service/vmdk_ops.py [hunk 7]
```diff
@@ -1071,7 +1084,8 @@ def handle_stale_attach(vmdk_path, kv_uuid):
           if cur_vm.runtime.powerState == VM_POWERED_OFF:
              logging.info("Detaching disk %s from VM(powered off) - %s\n",
                              vmdk_path, cur_vm.config.name)
-             device = findDeviceByPath(vmdk_path, cur_vm)
+             vol_name = vmdk_utils.strip_vmdk_extension(os.path.basename(vmdk_path))
```
Comment (msterin) [inline esx_service/vmdk_ops.py]: strip extension should be responsibility of the caller, but instead done in findDeviceByName (if needed)

File: esx_service/vmdk_ops.py (hunk 1 above)
Comment (shaominchen) [inline esx_service/vmdk_ops.py]: Filed issue #1070.

File: esx_service/vmdk_ops.py (hunk 6 above)
Comment (pshahzeb) [inline esx_service/vmdk_ops.py]: @msterin 
Fixing it to compare both volume name and datastore name with backing disk name and backing disk datastore.
Confirmed that we always get a qualified volume name i.e. volumename@datastore; from docker -> vmdk_plugin -> esx host.

File: esx_service/vmdk_ops.py (hunk 3 above)
Comment (pdhamdhere) [inline esx_service/vmdk_ops.py]: @pshahzeb ?

**Comment by pdhamdhere** [comment]
@pshahzeb Is it ready for review? Looks like you are still addressing review comments?

**Comment by pshahzeb** [comment]
Yes. This fix fails to detach volumes with two same short names but on different DS.  we get short name for volume for detach during volume create process.
Working on it.

File: esx_service/utils/vmdk_utils.py [hunk 8]
```diff
@@ -305,6 +309,28 @@ def get_vm_config_path(vm_name):
...
+        vm = vmdk_ops.findVmByUuid(vm_id)
+        for d in vm.config.hardware.device:
+            # if device is not a virtual disk, skip this device
+            if type(d) != vim.vm.device.VirtualDisk:
+                continue
+
+            # Filename format is as follows:
+            # "[<datastore name>] <parent-directory>/tenant/<vmdk-descriptor-name>"
+            # Trim the datastore name and keep disk path.
+            _, disk_path = d.backing.fileName.split(None, 1)
```
Comment (lipingxue) [inline esx_service/utils/vmdk_utils.py]: This cannot trim the datastore correctly if whitespaces are included in the datastore name.

>>> fileName = "[datastor e1] /dockvols/tenant1/vol1.vmdk"
>>> 
>>> _, disk_path = fileName.split(None, 1)
>>> print disk_path
e1] /dockvols/tenant1/vol1.vmdk

Use "]" as a separator.

File: esx_service/vmdk_ops.py [hunk 9]
```diff
@@ -1311,20 +1304,16 @@ def err(string):
     return {u'Error': string}
 
 
-def disk_detach(full_vol_name, vm):
-    """detach disk (by full_vol_name) from a vm and return None or err(msg)"""
+def disk_detach(vmdk_path, vm):
+    """detach disk (by full path) from a vm amd return None or err(msg)"""
```
Comment (lipingxue) [inline esx_service/vmdk_ops.py]: Typo: amd -> and

File: esx_service/utils/auth_api.py [hunk 10]
```diff
@@ -409,6 +409,17 @@ def _tenant_rm(name, remove_volumes=False):
         error_info = error_code.generate_error_info(ErrorCode.TENANT_NOT_EXIST, name)
         return error_info
 
+    # check if vms that are a part if this tenant have any volumes mounted.
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: "if this tenant" => "of this tenant"

File: esx_service/utils/auth_api.py [hunk 11]
```diff
@@ -409,6 +409,17 @@ def _tenant_rm(name, remove_volumes=False):
...
+    # If they have, can't delete the tenant.
+    if tenant.vms:
+        logging.info("tenant rm. VMs in tenant are %s", tenant.vms)
```
Comment (shaominchen) [inline esx_service/utils/auth_api.py]: Nit: "tenant rm" => "_tenant_rm", just for consistency.

File: esx_service/utils/error_code.py [hunk 12]
```diff
@@ -75,6 +76,7 @@ class ErrorCode:
     ErrorCode.VM_IN_ANOTHER_TENANT : "VM '{0}' has already been associated with tenant '{1}', can't add it",
     ErrorCode.VM_LIST_EMPTY : "VM list cannot be empty",
     ErrorCode.VM_DUPLICATE : "VMs {0} contain duplicates, they should be unique",
+    ErrorCode.VM_WITH_MOUNTED_VOLUMES :"VM '{0}' has volumes mounted.",
```
Comment (shaominchen) [inline esx_service/utils/error_code.py]: Nit: add a space after ":"

File: esx_service/utils/vmdk_utils.py [hunk 13]
```diff
@@ -307,6 +311,32 @@ def get_vm_config_path(vm_name):
     vm_config_path = os.path.join(datastore_path, path)
     return vm_config_path
 
+def check_volumes_mounted(vm_list):
+    """
+    Return error_info any vm in @param vm_list have docker volume mounted
```
Comment (shaominchen) [inline esx_service/utils/vmdk_utils.py]: Nit: add an "if" before "any vm"

File: esx_service/utils/vmdk_utils.py [hunk 14]
```diff
@@ -307,6 +311,32 @@ def get_vm_config_path(vm_name):
...
+                # if device is not a virtual disk, skip this device
+                if type(d) != vim.vm.device.VirtualDisk:
+                    continue
+
+                # Filename format is as follows:
+                # "[<datastore name>] <parent-directory>/tenant/<vmdk-descriptor-name>"
+                # Trim the datastore name and keep disk path.
+                _, disk_path = d.backing.fileName.rsplit("]", 1)
+                logging.info("backing disk name is %s", disk_path)
+                # If disk path start with "dockvols", the disk is a docker volume
```
Comment (shaominchen) [inline esx_service/utils/vmdk_utils.py]: Can we extract the logic of checking whether a disk is a docker volume (322-332) to a separate method?

**Pull request merged**